### PR TITLE
[FIX] *: read properties with the correct company

### PR DESCRIPTION
**Pull request by Feyensv**
All property_* fields grepped in the code, and if there is a company_id field on the current model, do the stuff in this company (self.env.company = company_id on the current model).

Some of those fixes should probably be backported if we think it is really needed in earlier versions.

Note that with_company was introduced in 13.1, thus earlier backports will have to replace `with_company` by `with_context(force_company=...**.id**)`

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/point_of_sale/models/pos_session.py [hunk 1]
```diff
@@ -667,7 +668,7 @@ def _prepare_line(self, order_line):
         """
         def get_income_account(order_line):
             product = order_line.product_id
-            income_account = product.with_company(order_line.company_id)._get_product_accounts()['income']
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_session.py]: Why do you want to remove this? It is wrong

Comment (Feyensv) [reply addons/point_of_sale/models/pos_session.py]: Be cause i put the with_company in a more global place, which means that get_income_account will always be called with the right company already.

Comment (pimodoo) [reply addons/point_of_sale/models/pos_session.py]: please don't change stuff like this in point_of_sale module without reasons, it doesn't give any improvement, and doesn't fix anything.

File: addons/stock/models/res_company.py [hunk 2]
```diff
@@ -124,34 +124,30 @@ def create_missing_warehouse(self):
     @api.model
     def create_missing_transit_location(self):
         company_without_transit = self.env['res.company'].search([('internal_transit_location_id', '=', False)])
-        for company in company_without_transit:
-            company._create_transit_location()
+        company_without_transit._create_transit_location()
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: doesn't seem related to this task

File: addons/stock/models/res_company.py [hunk 3]
```diff
@@ -124,34 +124,30 @@ def create_missing_warehouse(self):
...
 
     @api.model
     def create_missing_inventory_loss_location(self):
         company_ids  = self.env['res.company'].search([])
         inventory_loss_product_template_field = self.env['ir.model.fields'].search([('model','=','product.template'),('name','=','property_stock_inventory')])
         companies_having_property = self.env['ir.property'].search([('fields_id', '=', inventory_loss_product_template_field.id)]).mapped('company_id')
         company_without_property = company_ids - companies_having_property
-        for company in company_without_property:
-            company._create_inventory_loss_location()
+        company_without_property._create_inventory_loss_location()
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: doesn't seem related to this task

File: addons/stock/models/res_company.py [hunk 4]
```diff
@@ -124,34 +124,30 @@ def create_missing_warehouse(self):
...
 
     @api.model
     def create_missing_production_location(self):
         company_ids  = self.env['res.company'].search([])
         production_product_template_field = self.env['ir.model.fields'].search([('model','=','product.template'),('name','=','property_stock_production')])
         companies_having_property = self.env['ir.property'].search([('fields_id', '=', production_product_template_field.id)]).mapped('company_id')
         company_without_property = company_ids - companies_having_property
-        for company in company_without_property:
-            company._create_production_location()
+        company_without_property._create_production_location()
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: etc

File: addons/stock/models/res_company.py [hunk 5]
```diff
@@ -124,34 +124,30 @@ def create_missing_warehouse(self):
...
 
     @api.model
     def create_missing_scrap_location(self):
         company_ids  = self.env['res.company'].search([])
         companies_having_scrap_loc = self.env['stock.location'].search([('scrap_location', '=', True)]).mapped('company_id')
         company_without_property = company_ids - companies_having_scrap_loc
-        for company in company_without_property:
-            company._create_scrap_location()
+        company_without_property._create_scrap_location()
```
Comment (sle-odoo) [inline addons/stock/models/res_company.py]: etc

File: addons/stock/models/res_company.py (hunk 2 above)
Comment (Feyensv) [reply addons/stock/models/res_company.py]: Ok, will move it in another PR ;)

File: addons/point_of_sale/models/pos_session.py [hunk 6]
```diff
@@ -667,6 +668,7 @@ def _prepare_line(self, order_line):
         """
         def get_income_account(order_line):
             product = order_line.product_id
+            # VFE TODO this with_company can be dropped, it has no impact anymore.
```
Comment (Feyensv) [inline addons/point_of_sale/models/pos_session.py]: @pimodoo :stuck_out_tongue:

**Comment by MiquelRForgeFlow** [comment]
> Some of those fixes should probably be backported

I hope you do it at least for v13 🙏

**Comment by pimodoo** [comment]
@Feyensv maybe call it ref instead of FIX

**Comment by Feyensv** [comment]
I don't really get why it would be a ref :thinking:.
Ensuring all company_dependent fields are read with the correct company is a FIX for me ^^.

File: addons/stock_account/models/product.py [hunk 7]
```diff
@@ -460,6 +462,7 @@ def _svl_empty_stock_am(self, stock_valuation_layers):
         move_vals_list = []
         product_accounts = {product.id: product.product_tmpl_id.get_product_accounts() for product in stock_valuation_layers.mapped('product_id')}
         for out_stock_valuation_layer in stock_valuation_layers:
+            out_stock_valuation_layer = out_stock_valuation_layer.with_company(out_stock_valuation_layer.company_id)
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: _account_entry_move does the force_company/with_company so i'm not sure it's useful here?

File: addons/stock/models/stock_picking.py [hunk 8]
```diff
@@ -542,6 +542,7 @@ def _search_delay_alert_date(self, operator, value):
     @api.onchange('picking_type_id', 'partner_id')
     def onchange_picking_type(self):
         if self.picking_type_id:
+            self = self.with_company(self.company_id)
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: the company_id of a picking is a stored related to the company of the picking type, it seems to me this should be 
``` python
self = self.with_company(self.picking_type_id.company_id)
```

Comment (Feyensv) [reply addons/stock/models/stock_picking.py]: The compute of the related field should be triggered when reading it, but I can change it if you find it safer.

Comment (sle-odoo) [reply addons/stock/models/stock_picking.py]: ok just tested it and indeed self.company_id is the new one
i think it's better to be explicit, but up to you

File: addons/mrp/models/mrp_unbuild.py [hunk 9]
```diff
@@ -253,8 +253,10 @@ def _generate_move_from_existing_move(self, move, factor, location_id, location_
         })
 
     def _generate_move_from_bom_line(self, product, product_uom, quantity, bom_line_id=False, byproduct_id=False):
-        location_id = bom_line_id and product.property_stock_production or self.location_id
-        location_dest_id = bom_line_id and self.location_dest_id or product.with_company(self.company_id).property_stock_production
+        product_prod_location = product.with_company(self.company_id).property_stock_production
+        # VFE FIXME shouldn't it be self.location_id or product_prod_location ?
+        location_id = bom_line_id and product_prod_location or self.location_id
+        location_dest_id = bom_line_id and self.location_dest_id or product_prod_location
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_unbuild.py]: okay with the patch
about your comment, not having a bom_line_id means product is a finished produt -> should go from prod to the desired location, so the conditions are right

File: addons/purchase_stock/models/purchase.py [hunk 10]
```diff
@@ -187,7 +187,6 @@ def _get_picking_type(self, company_id):
             picking_type = self.env['stock.picking.type'].search([('code', '=', 'incoming'), ('warehouse_id', '=', False)])
         return picking_type[:1]
 
-    @api.model
```
Comment (sle-odoo) [inline addons/purchase_stock/models/purchase.py]: now i know that api model is completely useless -_-

File: addons/stock_account/models/stock_move.py [hunk 11]
```diff
@@ -411,6 +411,7 @@ def _generate_valuation_lines_data(self, partner_id, qty, debit_value, credit_va
 
         rslt = {'credit_line_vals': credit_line_vals, 'debit_line_vals': debit_line_vals}
         if credit_value != debit_value:
+            self = self.with_company(self.company_id)
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_move.py]: this one is useless i think (_account_entry_move does the with_company)

File: addons/stock_account/models/product.py (hunk 7 above)
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: probably useless

File: addons/stock_account/models/product.py [hunk 12]
```diff
@@ -96,6 +97,7 @@ class ProductProduct(models.Model):
     quantity_svl = fields.Float(compute='_compute_value_svl')
     stock_valuation_layer_ids = fields.One2many('stock.valuation.layer', 'product_id')
     valuation = fields.Selection(related="categ_id.property_valuation", readonly=True)
+    # VFE valuation is already defined on the template, what is the use of this field ?
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: i'm taking care of this removal in https://github.com/odoo/odoo/pull/45932

**Comment by Feyensv** [comment]
@robodoo rebase-ff r+
